ARIA cipher cmake

gojimmypi opened the pull request:
# Description

This is an initial  `cmake` PR for the MagicCrypto ARIA cipher suite as a supplement to https://github.com/wolfSSL/wolfssl/pull/6400/.

Currently the ARIA build automatically includes the MagicCrypto ciphers when the `ARIA_DIR` environment
variable is defined with a valid path to the directory structure extracted with `tar -xvf MagicCrypto.tar.gz`.

Reminder: Copy the contents of the `MagicCrypto` directory into the root of the wolfssl directory. Otherwise this error is observed during `./configure `

```
configure: error: Error including mcapi_error.h. Please put the MagicCrypto library in /mnt/c/test/wolfssl.
```

To use `cmake`, the `MagicCrypto` directory can be either in the wolfSSL tree, or specified with an environment variable: `ARIA_DIR`. 

Note this PR has since been updated to NOT enable ARIA just because an environment variable `ARIA_DIR` is defined. 

Enable ARIA with `-DWOLFSSL_ARIA=yes` cmake parameter.

### Build wolfSSL
For reference only, here is how to build using `make`
```bash
git clone https://github.com/wolfSSL/wolfssl.git
cd wolfssl
# copy MagicCrypto directory into wolfssl or set LD_LIBRARY_PATH:
# export LD_LIBRARY_PATH=/mnt/c/workspace/MagicCrypto/lib
./autogen.sh
./configure --enable-all --enable-aria
make
```


### Run CMake build
```bash
# set to your path
export ARIA_DIR=/mnt/c/workspace/MagicCrypto

mkdir -p out
pushd out
cmake .. -DWOLFSSL_ARIA=yes
cmake --build .

# View the available ciphers with:
./examples/client/client -e

# or with grep:
./examples/client/client -e | grep -i ARIA

popd
```

Fixes zd# n/a

# Testing

There's a [new ARIA CMake test script](https://github.com/gojimmypi/wolfssl/blob/aria/scripts/aria-cmake-build-test.sh) that checks:

* No ARIA Environment Variable, ARIA not enabled.
* No ARIA Environment Variable, ARIA Enabled
* ARIA Environment Variable with MagicCrypto in local user directory, ARIA not enabled.
* ARIA Environment Variable with MagicCrypto in local user directory, ARIA Enabled
* ARIA Environment Variable with MagicCrypto in wolfssl directory, ARIA not enabled
* ARIA Environment Variable with MagicCrypto in wolfssl, ARIA Enabled
* ARIA Environment Variable with bad directory, ARIA not enabled.
* ARIA Environment Variable with bad directory, ARIA Enabled (failure expected here)

### Windows Server Testing:

Unzip your `MagicCrypto.tar.gz`, shown here for `C:\workspace\MagicCrypto`

```bash
# set to your path
export ARIA_DIR=c:\\workspace\\MagicCrypto

mkdir -p out
pushd out
cmake .. -DWOLFSSL_ARIA=yes
cmake --build .

# View the available ciphers with:
./examples/client/client -e

popd
```

#### Linux Server Testing:

```bash
# From the root of the wolfSSL repo:

# set to your path
export ARIA_DIR=/mnt/c/workspace/MagicCrypto

mkdir -p out
pushd out
cmake .. -DWOLFSSL_ARIA=yes
cmake --build .

# View the available ciphers with:
./examples/client/client -e
popd
```

#### Server:

```bash
# reminder this is in the CMake `out` directory, NOT the root of wolfssl that is build with regular make.
# we also do NOT need the `export ARIA_DIR` here, as that's only for CMake building.

./examples/server/server -i -x -v 3 -A ./certs/ca-ecc-cert.pem -k ./certs/ecc-key.pem -c ./certs/intermediate/server-chain-ecc.pem -V -l ECDHE-ECDSA-ARIA128-GCM-SHA256
```

#### Client:

```bash
./examples/client/client -v 3 -l ECDHE-ECDSA-ARIA128-GCM-SHA256 -A ./certs/ca-ecc-cert.pem -k ./certs/ecc-client-key.pem -c ./certs/intermediate/client-chain-ecc.pem -C
```

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

bandi13 commented on test-aria.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
```
Not sure this file should be added to the repo.

gojimmypi replied:
agreed. removed

bandi13 reviewed (changes requested):
It's strange to me that you need to modify .C and .H files to get CMake to build with ARIA. Feels like something is not being set right.

dgarske commented on README_cmake_notes.md:
```diff
@@ -0,0 +1,27 @@
+cmake notes:
```
What is this file for?

gojimmypi replied:
nothing. the readme was only for interim reference. removed

dgarske commented on README-aria.md:
```diff
@@ -0,0 +1,71 @@
+# ARIA PR Notes:
```
Not sure we want another file in the root for this. If its is good info put in the normal README.md

gojimmypi replied:
agreed. will do

gojimmypi commented on examples/server/server.c:
```diff
@@ -109,6 +109,10 @@ static struct group_info group_id_to_text[] = {
 #endif
 
+#ifdef HAVE_ARIA
+    #include "aria-crypt.h"
```
yes, but without it the code would not compile. we certainly don't want it on all the time?

JacobBarthelmeh replied:
Still testing this out, but @gojimmypi can you elaborate on why it would not compile? Maybe slack me if the details are too long for here. This include and the additional ones added to aria-crypt.c and aria-cryptocb.c also caught my attention.

gojimmypi replied:
Hi @JacobBarthelmeh I must have been mistaken. I revisited this today & confirmed (twice) that indeed the include is not actually needed.

JacobBarthelmeh replied:
Checked the recent code, it looks like the include is still in multiple files. client.c for example.

gojimmypi replied:
oh yes, my bad. I had addressed only the `server.c` reference above. I've removed the include from the other examples.

I've also updated the PR description above with a bit more detail on building and testing. 

See also [the new README_cmake.md](https://github.com/gojimmypi/wolfssl/blob/aria/README_cmake.md) file.

gojimmypi commented on README-aria.md:
````diff
@@ -0,0 +1,71 @@
+
+```
+cd \workspace
+git clone https://github.com/gojimmypi/wolfssl.git wolfssl-gojimmypi-bandi13
````
the readme was only for interim reference. removed

gojimmypi commented on CMakeLists.txt:
```diff
@@ -2014,14 +2053,48 @@ set(LIB_SOURCES "")
 generate_lib_src_list("${LIB_SOURCES}")
+
 if(BUILD_SHARED_LIBS)
+    # message(STATUS "ARIA Check: SHARED! ${LIB_SOURCES}")
```
done

gojimmypi commented on CMakeLists.txt:
```diff
@@ -248,11 +248,13 @@ add_option("WOLFSSL_SINGLE_THREADED"
+        message(STATUS "ARIA Check: WOLFSSL_DEFINITIONS = ${WOLFSSL_DEFINITIONS}")
         list(APPEND WOLFSSL_DEFINITIONS
-            "-DHAVE_PTHREAD"
+        #    "-DHAVE_PTHREAD"
```
I believe this is resolved. I will pull the threading correction to a different PR.

gojimmypi commented on CMakeLists.txt:
```diff
@@ -18,8 +18,8 @@
 ####################################################
-
 cmake_minimum_required(VERSION 3.16)
+# set(CMAKE_BINARY_DIR "${CMAKE_BINARY_DIR}/build")
```
done

gojimmypi commented on CMakeLists.txt:
```diff
@@ -536,6 +538,40 @@ if(WOLFSSL_AESCTR AND NOT WOLFSSL_FORTRESS)
+    "no" "yes;no")
+
+if(0)
+        get_cmake_property(_variableNames VARIABLES)
```
done

gojimmypi commented on CMakeLists.txt:
```diff
@@ -536,6 +538,40 @@ if(WOLFSSL_AESCTR AND NOT WOLFSSL_FORTRESS)
+endif()
+
+if(WOLFSSL_ARIA)
+    message(STATUS "Have aria!")
```
agreed. removed

gojimmypi commented on CMakeLists.txt:
```diff
@@ -1882,6 +1918,9 @@ if (WOLFSSL_CAAM)
 endif()
 
+if (WOLFSSL_ARIA)
+   list(APPEND WOLFSSL_DEFINITIONS  "-DWOLFSSL_ARIA")
```
ooph, good catch!

gojimmypi commented:
I've removed the stray debug info files, added a couple of CMake README files, and put the CMake threading logic change in a separate PR (see https://github.com/wolfSSL/wolfssl/pull/6606).

I've confirmed this now works with Windows CMake (Visual Studio 2022 command prompt). For example the `server.exe` ends up in `out\examples\server\Debug`

```
C:\workspace\wolfssl-gojimmypi-bandi13\out>cmake ..
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22621.
-- The C compiler identification is MSVC 19.36.32535.0
-- The ASM compiler identification is MSVC
-- Found assembler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe
...etc
```

A separate issue that may need to be addressed: the Cmake-compiled examples for Windows do not immediately work fresh out of the box. 

Trying to run the executable, there's no error message, but nothing happens. This occurs when the wolfSSL dll's are not installed in the system. The current debug environment does not find them in the build directory. If I manually copy them, for example the 4 files created in the `\out\debug` directory:

```
         1,158,656 wolfssl.dll
            90,094 wolfssl.exp
           148,848 wolfssl.lib
         2,510,848 wolfssl.pdb
```

... to each of the respective example output directories: `out\examples\client\Debug` and `out\examples\server\Debug`. After this copy, then the Windows examples all work well. Perhaps it would be a good idea for CMake to copy them?

gojimmypi commented:
As noted in https://github.com/wolfSSL/wolfssl/pull/6606#issuecomment-1633334065 there may be a known problem in master unrelated to this PR. I've reverted the `WOLFSSL_SINGLE_THREADED` change here and squashed the commits.

The result is that all of the examples do not compile successfully with CMake in Windows for the ARIA ciphers. 

The test client and server _do_ compile successfully with CMake, in Windows and Linux, with and without the ARIA ciphers.

JacobBarthelmeh commented:
Please check the additional includes in other example files. For example client.c.

gojimmypi commented:
ok, I've removed all references of the ARIA `#include` that I added, retested everything, resync from upstream & pushed changes along with squash.

Please let me know if you see anything else that can be improved.

gojimmypi commented:
I just noticed that @dgarske [suggested](https://github.com/wolfSSL/wolfssl/pull/6600#discussion_r1261645083) I not add the new [README_cmake.md](https://github.com/gojimmypi/wolfssl/blob/aria/README_cmake.md) to the root. I've not yet merged that into the main [README.md](https://github.com/wolfSSL/wolfssl/blob/master/README.md) file.

I'm not sure this even belongs in the main `README.md` file, perhaps best to just move it into the [cmake directory](https://github.com/wolfSSL/wolfssl/tree/master/cmake), eh? Note that the `cmake` directory is not the main location of CMake files, rather only supplementary function as noted in my new [README.md](https://github.com/gojimmypi/wolfssl/blob/aria/cmake/README.md) file there.

I'll await any further code review comments before making additional changes. @JacobBarthelmeh let me know what you prefer. Thanks

bandi13 commented on CMakeLists.txt:
```diff
@@ -2001,13 +2020,45 @@ set(LIB_SOURCES "")
+    if (ARIA_FOUND)
+        # message(STATUS "find_package FOUND ARIA")
+    else()
+        message(WARNING "WOLFSSL_ARIA is enabled, but find_package did not find ARIA MagicCrypto")
```
This should probably be an error

gojimmypi replied:
Agreed

bandi13 commented on CMakeLists.txt:
```diff
@@ -531,6 +531,22 @@ if(WOLFSSL_AESCTR AND NOT WOLFSSL_FORTRESS)
 endif()
 
+# if we find an ARIA_DIR turn on ARIA support
+if ("$ENV{ARIA_DIR}" STREQUAL "")
```
This check is kind of tricky. What if someone is trying to build without ARIA, doesn't supply the WOLFSSL_ARIA flag, but forgets about ARIA_DIR? I don't think having such an environment variable set should automatically turn on ARIA.

gojimmypi replied:
My original plan was to turn on ARIA feature when the `ARIA_DIR` environment variable was found. I can see how that might be undesired.

I have a new solution in the works that will ignore the  `ARIA_DIR` if not enabled. Now the only way to enable is via Cmake flags, such as ` cmake .. -DWOLFSSL_ARIA=yes`

PR update soon.

bandi13 commented on cmake/modules/FindARIA.cmake:
```diff
@@ -0,0 +1,70 @@
+
+# Make sure we have a ARIA_DIR encironment variable with the path to MagicCrypto
+if ("$ENV{ARIA_DIR}" STREQUAL "")
+    message(ERROR "ERROR: FindARIA.cmake missing ARIA_DIR value")
```
I think this should default to ARIA_DIR=WOLFSSL_ROOT/MagicCrypto instead of giving an error. With autotools, MagicCrypto is explicitly required to be in the root of the wolfSSL directory. We should keep the two build systems in sync. Either have CMake require it in the same spot, or update configure.ac to support the ARIA_DIR. My vote is for the former.

gojimmypi replied:
I've added additional logic here for _either_ the environment variable _or_ a local copy of MagicCrypto in the wolfSSL repo directory structure.  I've added a new `ARIA_IS_LOCAL` variable and will update this PR description, above. 

Easier said than done, as `target_include_directories()` complains unintuitively if the respective directory is in the current cmake directory structure. See [stackoverflow](https://stackoverflow.com/questions/25676277/cmake-target-include-directories-prints-an-error-when-i-try-to-add-the-source) and the [docs](https://cmake.org/cmake/help/latest/command/target_include_directories.html).

PR update soon.

gojimmypi replied:
See the [scripts/aria_cmake_build_test.sh](https://github.com/gojimmypi/wolfssl/blob/aria/scripts/aria_cmake_build_test.sh)

JacobBarthelmeh replied:
I like the idea of having the option to set the ARIA directory. Agreed though that it should default to checking the wolfSSL root directory to keep the behavior the same if ARIA_DIR is unset.

bandi13 commented on cmake/modules/FindARIA.cmake:
```diff
@@ -0,0 +1,70 @@
+# Check that the appropriate files exist
+find_path(ARIA_INCLUDE_DIR NAMES "mcapi.h" )
+
+if (EXISTS "${ARIA_INCLUDE_DIR}/mcapi.h")
```
Negate these three checks to simplify and remove the commented out block.

gojimmypi replied:
logic is now "if not exists.... error"

bandi13 commented on wolfcrypt/src/port/aria/aria-crypt.c:
```diff
@@ -38,6 +38,8 @@ size and a key size of 128, 192, or 256 bits.
 #include <wolfssl/wolfcrypt/error-crypt.h>
 #include <wolfssl/ssl.h>
 #include <wolfssl/wolfcrypt/port/aria/aria-crypt.h>
+#include <wolfssl/wolfcrypt/aes.h>
```
It's unusual that this should still be necessary. Have you looked to see what is the difference in Make vs CMake and why the latter needs this?

gojimmypi replied:
I'll see what I can find. 

Without that `aes.h` include, I see this error when running `cmake --build .`

```
Consolidate compiler generated dependencies of target wolfssl
[  1%] Building C object CMakeFiles/wolfssl.dir/wolfcrypt/src/port/aria/aria-crypt.c.o
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c: In function ‘CheckAriaGcmIvSize’:
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c:144:21: error: ‘GCM_NONCE_MIN_SZ’ undeclared (first use in this function)
  144 |     return (ivSz == GCM_NONCE_MIN_SZ ||
      |                     ^~~~~~~~~~~~~~~~
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c:144:21: note: each undeclared identifier is reported only once for each function it appears in
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c:145:21: error: ‘GCM_NONCE_MID_SZ’ undeclared (first use in this function)
  145 |             ivSz == GCM_NONCE_MID_SZ ||
      |                     ^~~~~~~~~~~~~~~~
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c:146:21: error: ‘GCM_NONCE_MAX_SZ’ undeclared (first use in this function)
  146 |             ivSz == GCM_NONCE_MAX_SZ);
      |                     ^~~~~~~~~~~~~~~~
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c: In function ‘wc_AriaGcmSetIV’:
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c:175:42: error: ‘AES_IV_FIXED_SZ’ undeclared (first use in this function)
  175 |         (ivFixed != NULL && ivFixedSz != AES_IV_FIXED_SZ)) {
      |                                          ^~~~~~~~~~~~~~~
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c: In function ‘CheckAriaGcmIvSize’:
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-crypt.c:147:1: error: control reaches end of non-void function [-Werror=return-type]
  147 | }
      | ^
cc1: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/wolfssl.dir/build.make:286: CMakeFiles/wolfssl.dir/wolfcrypt/src/port/aria/aria-crypt.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:101: CMakeFiles/wolfssl.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
gojimmypi:/mnt/c/test/wolfssl-PR-aria/out
```

JacobBarthelmeh replied:
Similar chain of includes as ecc. aria-cryptocb.h -> cryptocb.h -> aes.h

gojimmypi replied:
@JacobBarthelmeh AND @bandi13 I've confirmed that turning on `OPENSSL_EXTRA` negates the need for the `#include <wolfssl/wolfcrypt/ecc.h>` in `aria-cryptocb.c` and for the `#include <wolfssl/wolfcrypt/aes.h>` & `#include <wolfssl/error-ssl.h>` in  `aria-crypt.c`.

```
cmake ..  -DWOLFSSL_ARIA=yes -DWOLFSSL_OPENSSLEXTRA=yes
```

I've also confirmed the client & server apps work properly without  `OPENSSL_EXTRA` defined.

This also applies to regular `make` build:

```

./configure --enable-aria --disable-opensslall --disable-opensslextra
make
```

Thanks @bandi13 for all the offline comments & suggestions.

I've created https://github.com/wolfSSL/wolfssl/pull/6730 that adds some wolfSSL includes to ARIA and removes the dependency on OpenSSL needing to be enabled to use ARIA.

bandi13 commented on wolfcrypt/src/port/aria/aria-cryptocb.c:
```diff
@@ -39,6 +39,7 @@ size and a key size of 128, 192, or 256 bits.
 #include <wolfssl/ssl.h>
 #include <wolfssl/wolfcrypt/asn_public.h>
 #include <wolfssl/wolfcrypt/port/aria/aria-cryptocb.h>
+#include <wolfssl/wolfcrypt/ecc.h>
```
Same question here as in `aria-crypt.c`

gojimmypi replied:
The `ecc_set_type` used in [cryptocb.c](https://github.com/wolfSSL/wolfssl/blob/827287000c7f075eff0708f0b8839c778f10fd67/wolfcrypt/src/port/aria/aria-cryptocb.c#L182) is [defined in ecc.h](https://github.com/wolfSSL/wolfssl/blob/827287000c7f075eff0708f0b8839c778f10fd67/wolfssl/wolfcrypt/ecc.h#L289) 

I suspect a setting in the regular `make` causes it to be already included.

Omitting that `#include` for `cmake` gives these build errors:

```
[  1%] Building C object CMakeFiles/wolfssl.dir/wolfcrypt/src/port/aria/aria-cryptocb.c.o
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-cryptocb.c: In function ‘wc_AriaSign’:
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-cryptocb.c:183:11: error: unknown type name ‘ecc_set_type’
  183 |     const ecc_set_type* dp;
      |           ^~~~~~~~~~~~
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-cryptocb.c:217:12: error: invalid use of incomplete typedef ‘ecc_key’
  217 |     if (key->dp != NULL) {
      |            ^~
/mnt/c/test/wolfssl-PR-aria/wolfcrypt/src/port/aria/aria-cryptocb.c:218:17: error: invalid use of incomplete typedef ‘ecc_key’
  218 |         dp = key->dp;
      |                 ^~
[... snip ... ]

many other similar errors....
```

JacobBarthelmeh replied:
The chain of includes should go wolfssl/wolfcrypt/port/aria/aria-cryptocb.h -> wolfssl/wolfcrypt/cryptocb.h -> wolfssl/wolfcrypt/ecc.h.

Possible reasons for it to go awry would be WOLF_CRYPTO_CB not defined or if this is on windows then maybe some interaction with the WOLFSSL_ECC_CURVE_STATIC define which gets set on Windows and slightly changes the ecc_set_type struct?

gojimmypi replied:
Hi @JacobBarthelmeh - do you prefer that I include `cryptocb.h` _instead_ of `ecc.h` in my [proposed changes](https://github.com/wolfSSL/wolfssl/pull/6730/files)?

Doing so does not work by default. I see a ton of errors. 

I tried to force ECC on by adding this to the beginning of the `aria-crypt.c` and `aria-cryptcb.c` files:

```
#undef HAVE_ECC
#define HAVE_ECC
```

For example:
```
wolfcrypt/src/port/aria/aria-cryptocb.c:261:11: error: unknown type name ‘ecc_set_type’
  261 |     const ecc_set_type* dp;
      |           ^~~~~~~~~~~~
wolfcrypt/src/port/aria/aria-cryptocb.c:297:12: error: invalid use of incomplete typedef ‘ecc_key’
  297 |     if (key->dp != NULL) {
      |            ^~
wolfcrypt/src/port/aria/aria-cryptocb.c:298:17: error: invalid use of incomplete typedef ‘ecc_key’
  298 |         dp = key->dp;
      |                 ^~
wolfcrypt/src/port/aria/aria-cryptocb.c:301:41: error: invalid use of incomplete typedef ‘ecc_key’
  301 |         dp = wc_ecc_get_curve_params(key->idx);
      |                                         ^~
wolfcrypt/src/port/aria/aria-cryptocb.c:304:11: error: request for member ‘id’ in something not a structure or union
  304 |     if (dp->id == ECC_SECP256R1) {
```

I could try and chase this one to see what directives need to be turned on, but it seems considerably more straightforward to simply include the needed `wolfcrypt/ecc.h` file.

Note with the additional ARIA includes and the changes in https://github.com/wolfSSL/wolfssl/pull/6730, the dependency on having OpenSSL enabled has been removed.

Let me know which route your prefer. Thanks.

gojimmypi commented on CMakeLists.txt:
```diff
@@ -2282,6 +2333,15 @@ else()
   endif()
 endif()
 
+if(NOT BUILD_ARIA)
```
This was only following the adjacent pattern of "_if files found but not needed for this build, ignore them_".

Still, unless there's an explicit `include` for the MagicCrypto, the files would not be found anyhow. I'll be removing this section,

gojimmypi commented:
I've made some cmake changes & updated the PR description.

gojimmypi commented:
For reference, in the context of having a wolfssl directory tree containing the MagicCrypto directory, there are known problems on WSL / Windows symlinks. 

See https://github.com/microsoft/WSL/issues/5118 and [various others](https://github.com/microsoft/WSL/issues?q=is%3Aissue+is%3Aopen+symlink).

JacobBarthelmeh commented on README_cmake.md:
```diff
@@ -0,0 +1,105 @@
+# wolfSSL CMake notes
```
Good notes! Please incorporate them into the INSTALL file under the "Building with CMake" section and remove README_cmake.md. If it gets to lengthy in the INSTALL the additional details can be in the wolfSSL Manual.

gojimmypi replied:
ok, `README_cmake.md` removed and text added to the `INSTALL` file.

dgarske commented on cmake/README.md:
```diff
@@ -0,0 +1,6 @@
+
+This directory contains some supplementary functions for the [CMakeLists.txt](../CMakeLists.txt) in the root.
+
+See also the [README_cmake.md](../README_cmake.md) file.
```
Fix. No longer README_cmake.md.

gojimmypi replied:
I updated that to instead reference the root `INSTALL` document.

dgarske reviewed (changes requested):
Works as described. Just a few minor documentation fixes needed.

gojimmypi commented on INSTALL:
```diff
@@ -110,6 +110,55 @@
 
+    In the simplest form:
+
+    # create a root directory for wolfssl repo
```
ah yes, good. I removed those lines from the docs.

gojimmypi commented:
All checks pass. @dgarske let me know if you'd prefer I squash the 14 commits in this PR.

dgarske commented:
> All checks pass. @dgarske let me know if you'd prefer I squash the 14 commits in this PR.

I squashed on merge. Thanks
